File: src/etch/bindings/java/support/Validator_string.java
package etch.bindings.java.support;

import etch.bindings.java.msg.Validator;

/**
 * Validator for string.
 */
public class Validator_string extends TypeValidator
{
	/**
	 * @param nDims the number of dimensions. 0 for a scalar.
	 * @return an instance of the validator.
	 */
	public static Validator_string get( int nDims )
	{
		checkDims( nDims );
		
		if (nDims >= validators.length)
			return new Validator_string( nDims );
		
		Validator_string v = validators[nDims];
		
		if (v == null)
			v = validators[nDims] = new Validator_string( nDims );
		
		return v;
	}
	
	private static Validator_string[] validators = new Validator_string[MAX_CACHED];
	
	/**
	 * Constructs the Validator.
	 *
	 * @param nDims the number of dimensions. 0 for a scalar.
	 */
	private Validator_string( int nDims )
	{
		super( String.class, String.class, nDims, "string["+nDims+"]" );
	}

	@Override
	public Validator elementValidator()
	{
		return get( nDims-1 );
	}
}

File: binding-python/runtime/src/main/python/etch/binding/support/Validator_string.py
"""
#  Licensed to the Apache Software Foundation (ASF) under one   *
#  or more contributor license agreements.  See the NOTICE file *
#  distributed with this work for additional information        *
#  regarding copyright ownership.  The ASF licenses this file   *
#  to you under the Apache License, Version 2.0 (the            *
#  "License"); you may not use this file except in compliance   *
#  with the License.  You may obtain a copy of the License at   *
#                                                               *
#    http://www.apache.org/licenses/LICENSE-2.0                 *
#                                                               *
#  Unless required by applicable law or agreed to in writing,   *
#  software distributed under the License is distributed on an  *
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
#  KIND, either express or implied.  See the License for the    *
#  specific language governing permissions and limitations      *
#  under the License. 
"""
from __future__ import absolute_import
import types

from ...python.Exceptions import *
from ...python.Types import *
from .TypeValidator import *
from ..transport.fmt.TypeCode import *


class Validator_string(ScalarAndArrayTypeValidator):

    def __init__(self, nDims):
        """
        Construct the validator.

        @param nDims    the number of dimensions. 0 or scalar.
        """
        typs = [String,types.StringType,types.UnicodeType]
        super(Validator_string,self).__init__(typs, None, nDims, "string[%d]" % nDims, False, TypeCode.STRING)

    def checkValue(self, value):
        tc = super(Validator_string,self).checkValue(value)

        # Override with EMPTY_STRING if len(value) == 0
        if tc == TypeCode.STRING:
            if len(value) == 0:
                tc = TypeCode.EMPTY_STRING
        return tc
